Update equalLettersIgnoringASCIICase() to take in a span

cdumez opened the pull request:
#### 28b97118a65d3b803043588d385571dddf4eaf70
<pre>
Update equalLettersIgnoringASCIICase() to take in a span
<a href="https://bugs.webkit.org/show_bug.cgi?id=274085">https://bugs.webkit.org/show_bug.cgi?id=274085</a>

Reviewed by Darin Adler.

* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::parse):
* Source/WTF/wtf/DateMath.cpp:
(WTF::parseDate):
* Source/WTF/wtf/text/StringCommon.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::messageSourceForWTFLogChannel):
* Source/WebCore/html/parser/ParsingUtilities.h:
(WebCore::skipExactlyIgnoringASCIICase):

Canonical link: <a href="https://commits.webkit.org/278730@main">https://commits.webkit.org/278730@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/623f2d9142201eb30a8a534ac19924c690661df7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51408 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30718 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3759 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54675 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2101 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/53711 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37036 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1781 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/41867 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53507 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28362 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44337 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22990 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/25672 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1596 "Passed tests") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/44751 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47639 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1673 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56267 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/50912 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26527 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1557 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/49272 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27767 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44397 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/48454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28660 "Built successfully") | | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/63234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7490 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27502 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/63234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/graphics/gstreamer/DMABufVideoSinkGStreamer.cpp:
```diff
@@ -180,7 +180,8 @@ bool webKitDMABufVideoSinkIsEnabled()
     std::call_once(s_flag, [&] {
         const char* value = g_getenv("WEBKIT_GST_DMABUF_SINK_DISABLED");
-        s_disabled = value && (equalLettersIgnoringASCIICase(value, "true"_s) || equalLettersIgnoringASCIICase(value, "1"_s));
+        auto valueSpan = span(value);
```
Does this `span` function work on null pointers?

cdumez replied:
Yes it does:
```
inline std::span<const LChar> span8(const char* string)
{
    return { reinterpret_cast<const LChar*>(string), string ? strlen(string) : 0 };
}
```

darinadler commented on Source/WTF/wtf/DateMath.cpp:
```diff
@@ -960,7 +960,7 @@ double parseDate(std::span<const LChar> dateString, bool& isLocalTime)
                 // dateString has the time zone name as a prefix, not that it is equal.
                 auto tzName = span8(knownZone.tzName);
-                if (dateString.size() >= tzName.size() && equalLettersIgnoringASCIICase(dateString.data(), tzName)) {
+                if (dateString.size() >= tzName.size() && equalLettersIgnoringASCIICaseWithLength(dateString, tzName, tzName.size())) {
```
Could this use `dateString.first(tzName.size())` instead of calling a different function?

darinadler commented on Source/WebCore/html/parser/ParsingUtilities.h:
```diff
@@ -213,7 +213,7 @@ template<typename CharacterType> bool skipExactlyIgnoringASCIICase(const Charact
     if (position + literalLength > end)
         return false;
-    if (!equalLettersIgnoringASCIICase(position, literal))
+    if (!equalLettersIgnoringASCIICaseWithLength(std::span { position, literalLength }, literal.span8(), literalLength))
```
I don’t understand why `literalLength` has to be passed here.

cdumez replied:
If we call equalLettersIgnoringASCIICase() instead of equalLettersIgnoringASCIICaseWithLength(), it will check that the lengths of the 2 parameters are equal. This check would be redundant here.

darinadler commented on Source/WebCore/html/parser/ParsingUtilities.h:
```diff
@@ -225,7 +225,7 @@ template<typename CharacterType> bool skipExactlyIgnoringASCIICase(StringParsing
     if (buffer.lengthRemaining() < literalLength)
         return false;
-    if (!equalLettersIgnoringASCIICase(buffer.position(), literal))
+    if (!equalLettersIgnoringASCIICaseWithLength(buffer.span(), literal.span8(), literalLength))
```
Can this use `buffer.span().first(lliteralLength)`?

cdumez replied:
Are you arguing for calling equalLettersIgnoringASCIICase() instead of equalLettersIgnoringASCIICaseWithLength() by taking sure the first 2 parameters have the same length?

The issue is that this wouldn't be as efficient:
- There would be the extra cost of calling `first(literalLength)` on the first parameter
- equalLettersIgnoringASCIICase() would start by making sure that the 2 input spans have the same length (which would be redundant).

FYI, the pattern I used in this PR is that same one I had used in my earlier PR for equalIgnoringASCIICase() + equalIgnoringASCIICaseWithLength().

darinadler replied:
OK

cdumez committed:
Update equalLettersIgnoringASCIICase() to take in a span
https://bugs.webkit.org/show_bug.cgi?id=274085

Reviewed by Darin Adler.

* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::parse):
* Source/WTF/wtf/DateMath.cpp:
(WTF::parseDate):
* Source/WTF/wtf/text/StringCommon.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::messageSourceForWTFLogChannel):
* Source/WebCore/html/parser/ParsingUtilities.h:
(WebCore::skipExactlyIgnoringASCIICase):

Canonical link: https://commits.webkit.org/278730@main
